# Homework Assignment - Lesson 6: Combining Datasets - Joins

**Due Date:** [10/12/2025]

**Instructions:**
- Complete the following tasks in this R notebook
- Use the exact variable names specified for each task
- Ensure your code runs without errors and produces the expected outputs
- Submit your completed notebook with all code and outputs

---

## Part 1: Data Import and Setup

**Task:** Import the following CSV files and examine their structure.

**Required Variable Names:**
You must use these EXACT variable names:
- `customers` for customers.csv
- `orders` for orders.csv  
- `order_items` for order_items.csv
- `products` for products.csv
- `suppliers` for suppliers.csv

**Expected Outputs:**
- All datasets loaded successfully
- Print dataset dimensions using `nrow()` and `ncol()`
- Show first 3 rows of each dataset using `head(dataset, 3)`

In [53]:
# Your code here:
# 1. Load the tidyverse package
# 2. Set working directory to sample_datasets  
# 3. Import all CSV files using read_csv() with EXACT variable names above
# 4. Print dimensions and head() for each dataset

# Load tidyverse

library(tidyverse)

# Set working directory
setwd("/workspaces/assignment-4-2-/data")  

# Import datasets with exact variable names
customers <- read_csv("customers.csv")
orders <- read_csv("orders.csv")
order_items <- read_csv("order_items.csv")
products <- read_csv("products.csv")
suppliers <- read_csv("suppliers.csv")

# Print dataset information (uncomment after creating variables above):
# cat("Dataset Dimensions:\n")
# cat("Customers:", nrow(customers), "rows x", ncol(customers), "columns\n")
# cat("Orders:", nrow(orders), "rows x", ncol(orders), "columns\n")
# cat("Order Items:", nrow(order_items), "rows x", ncol(order_items), "columns\n")
# cat("Products:", nrow(products), "rows x", ncol(products), "columns\n")
# cat("Suppliers:", nrow(suppliers), "rows x", ncol(suppliers), "columns\n")

# Show first 3 rows of each dataset (uncomment after loading data):
# head(customers, 3)
# head(orders, 3)
# head(order_items, 3)
# head(products, 3)
# head(suppliers, 3)

Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (3): Name, Email, City
dbl  (1): CustomerID
date (1): Registration_Date



ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 250 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl  (3): OrderID, CustomerID, Total_Amount
date (1): Order_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 400 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
dbl (4): OrderID, ProductID, Quantity, Unit_Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 50 Columns: 4
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (2): Product_Name, Category
dbl (2): ProductID, Supplier_ID

ℹ Use

## Part 2: Basic Joins

**Tasks:**
1. **Inner Join:** Create `customer_orders_inner` by joining customers and orders
2. **Left Join:** Create `customer_orders_left` to show all customers  
3. **Right Join:** Create `customer_orders_right` to show all orders
4. **Full Join:** Create `customer_orders_full` for complete data view

Use the exact variable names specified above for autograding. Analyze row counts and explain results.

In [54]:
# Your code here:
# Create an inner join between customers and orders
# Analyze the results and compare row counts
customer_orders <- inner_join(customers, orders, by = "CustomerID")
cat("Customer Orders Dimensions:", nrow(customer_orders), "rows x", ncol(customer_orders), "columns\n")


Customer Orders Dimensions: 200 rows x 8 columns


In [55]:
# Your code here:
# Create a left join to keep all customers
# Count how many customers have no orders
customer_orders_left <- left_join(customers, orders, by = "CustomerID")
no_order_count <- sum(is.na(customer_orders_left$OrderID))
cat("Customers with No Orders:", no_order_count, "\n")


Customers with No Orders: 0 


In [56]:
# Your code here:
# Create a right join to keep all orders
# Check for orders with invalid customer IDs
customer_orders_right <- right_join(customers, orders, by = "CustomerID")
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Orders with Invalid Customer IDs:", orders_invalid_customers, "\n")


Orders with Invalid Customer IDs: 50 


In [57]:
## Part 2: Basic Joins

**Required Variable Names:**
You must create these EXACT variable names:

Part 2.1:** Create `customer_orders` using inner_join()

Part 2.2:** Create `customer_orders_left` using left_join() 

Part 2.3:** Create `customer_orders_right` using right_join()

Part 2.4:** Create `customer_orders_full` using full_join()

**Expected Outputs:**
For each join, print:
- Number of rows in the result using `nrow()`
- Comparison with original dataset sizes
- Analysis of unmatched records

ERROR: Error in parse(text = input): <text>:3:1: unexpected '^'
2: 
3: **
   ^


In [ ]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- inner_join(customers, orders, by = "CustomerID")
cat("Inner Join - Customer Orders Rows:", nrow(customer_orders), "\n")

# Required output for autograding:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join - Customer Orders Rows: 200 
Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [ ]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- left_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [ ]:
# Part 2.1: Inner Join (REQUIRED variable name: customer_orders)
# Your code here:
customer_orders <- inner_join(customers, orders, by = "CustomerID")
cat("Inner Join - Customer Orders Rows:", nrow(customer_orders), "\n")

# Expected output:
cat("Inner Join Results:
")
cat("Customers:", nrow(customers), "
")
cat("Orders:", nrow(orders), "
")
cat("Inner Join Result:", nrow(customer_orders), "
")

Inner Join - Customer Orders Rows: 200 
Inner Join Results:
Customers: 100 
Orders: 250 
Inner Join Result: 200 


In [ ]:
# Part 2.2: Left Join (REQUIRED variable name: customer_orders_left)  
# Your code here:
customer_orders_left <- left_join(customers, orders, by = "CustomerID")

# Expected output:
customers_without_orders <- sum(is.na(customer_orders_left$OrderID))
cat("Left Join Results:
")
cat("Total rows:", nrow(customer_orders_left), "
")
cat("Customers without orders:", customers_without_orders, "
")

Left Join Results:
Total rows: 200 
Customers without orders: 0 


In [ ]:
# Part 2.3: Right Join (REQUIRED variable name: customer_orders_right)
# Your code here:
customer_orders_right <- right_join(customers, orders, by = "CustomerID")

# Expected output:
orders_invalid_customers <- sum(is.na(customer_orders_right$Name))
cat("Right Join Results:
")
cat("Total rows:", nrow(customer_orders_right), "
")
cat("Orders with invalid customer IDs:", orders_invalid_customers, "
")

Right Join Results:
Total rows: 250 
Orders with invalid customer IDs: 50 


In [ ]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- full_join(customers, orders, by = "CustomerID")

# Expected output:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


In [ ]:
# Part 2.4: Full Join (REQUIRED variable name: customer_orders_full)
# Your code here:
customer_orders_full <- full_join(customers, orders, by = "CustomerID")

# Required output for autograding:
customers_only <- sum(is.na(customer_orders_full$OrderID))
orders_only <- sum(is.na(customer_orders_full$Name))
cat("Full Join Results:
")
cat("Total rows:", nrow(customer_orders_full), "
")
cat("Customers without orders:", customers_only, "
")
cat("Orders without valid customers:", orders_only, "
")

Full Join Results:
Total rows: 250 
Customers without orders: 0 
Orders without valid customers: 50 


## Part 3: Multi-Table Joins

**Tasks:**
1. Create a comprehensive dataset by joining `orders`, `customers`, and `order_items`
2. Extend this dataset by adding `products` and `suppliers` information
3. Create a complete supply chain view showing the full customer-to-supplier relationship

Build your joins step by step and examine the results at each stage.

## Part 3: Multi-Table Joins

**Requirements for Autograding:**
You must create these EXACT variable names in order:

1. `orders_items` - Join orders and order_items  
2. `orders_customers_items` - Add customers to above
3. `complete_order_data` - Add products information
4. `complete_data` - Add suppliers for complete supply chain view

**Expected Outputs:**
Print the number of rows for each step to show the join progression.

In [ ]:
# Step 1: REQUIRED variable name: orders_items
# Your code here:
orders_items <- inner_join(orders, order_items, by = "OrderID")

# Required output for autograding:
cat("Step 1 - Orders with Items:", nrow(orders_items), "rows\n")

Step 1 - Orders with Items: 400 rows


In [ ]:
# Step 2: REQUIRED variable name: orders_customers_items  
# Your code here:
orders_customers_items <- inner_join(orders_items, customers, by = "CustomerID")

# Required output for autograding:
cat("Step 2 - Add Customers:", nrow(orders_customers_items), "rows\n")

Step 2 - Add Customers: 310 rows


In [ ]:
# Step 3: REQUIRED variable name: complete_order_data
# Your code here:
complete_order_data <- inner_join(orders_customers_items, products, by = "ProductID")

# Required output for autograding:
cat("Step 3 - Add Products:", nrow(complete_order_data), "rows\n")

Step 3 - Add Products: 310 rows


In [ ]:
# Step 4: REQUIRED variable name: complete_data
# Your code here:
complete_data <- inner_join(complete_order_data, suppliers, by = "Supplier_ID")

# Required output for autograding:
cat("Step 4 - Complete Supply Chain:", nrow(complete_data), "rows\n")

Step 4 - Complete Supply Chain: 310 rows


## Part 4: Data Quality Analysis

**Tasks:**
1. Use `anti_join()` to find unmatched records between tables
2. Use `semi_join()` to find matched records
3. Check for duplicate keys and analyze their impact on joins
4. Identify and document any data quality issues

This analysis helps understand data integrity and potential issues in your datasets.

In [ ]:
<VSCode.Cell id="#VSC-d7f91846" language="markdown">
## Part 4: Data Quality Analysis

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customers_no_orders` - Use anti_join() to find customers with no orders
2. `orphaned_orders` - Use anti_join() to find orders without customers  
3. `products_never_ordered` - Use anti_join() to find unordered products
4. `active_customers` - Use semi_join() to find customers with orders

**Expected Outputs:**
Print the count of records for each using `nrow()`.

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [ ]:
# REQUIRED variable name: customers_no_orders
# Your code here:
customers_no_orders <- anti_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who never placed an order:", nrow(customers_no_orders), "\n")

Customers who never placed an order: 0 


In [ ]:
# REQUIRED variable name: orphaned_orders
# Your code here:
orphaned_orders <- anti_join(orders, customers, by = "CustomerID")

# Required output for autograding:
cat("Orders without corresponding customers:", nrow(orphaned_orders), "\n")

Orders without corresponding customers: 50 


In [ ]:
# REQUIRED variable name: products_never_ordered
# Your code here:
products_never_ordered <- anti_join(products, order_items, by = "ProductID")

# Required output for autograding:
cat("Products that were never ordered:", nrow(products_never_ordered), "\n")

Products that were never ordered: 0 


In [ ]:
# REQUIRED variable name: active_customers
# Your code here:
active_customers <- semi_join(customers, orders, by = "CustomerID")

# Required output for autograding:
cat("Customers who placed at least one order:", nrow(active_customers), "\n")

Customers who placed at least one order: 100 


In [ ]:
# Data integrity checks - count duplicate keys
# Your code here: Calculate and print duplicate counts for each dataset

# Required outputs for autograding (uncomment after creating the calculations above):
# duplicate_customers <- sum(duplicated(customers$CustomerID))
# duplicate_orders <- sum(duplicated(orders$OrderID))
# duplicate_products <- sum(duplicated(products$ProductID))
# duplicate_suppliers <- sum(duplicated(suppliers$Supplier_ID))

# cat("Data Quality Summary:\n")
# cat("Duplicate customer IDs:", duplicate_customers, "\n")
# cat("Duplicate order IDs:", duplicate_orders, "\n")
# cat("Duplicate product IDs:", duplicate_products, "\n")
# cat("Duplicate supplier IDs:", duplicate_suppliers, "\n")

## Part 5: Business Analysis with Joined Data

**Tasks:**
1. Calculate customer lifetime value metrics (total spent, number of orders, average order value)
2. Analyze product performance (total quantity sold, revenue generated, order frequency)
3. Evaluate supplier performance (total revenue, number of products supplied)
4. Create a regional analysis by customer city

Use your joined datasets to generate meaningful business insights.

In [ ]:
<VSCode.Cell id="#VSC-f7364f43" language="markdown">
## Part 5: Business Analysis with Joined Data

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `customer_metrics` - Customer analysis with Total_Spent, Order_Count, Avg_Order_Value
2. `product_metrics` - Product analysis with Total_Revenue, Total_Quantity, Order_Frequency  
3. `supplier_metrics` - Supplier analysis with Product_Count by supplier
4. `regional_analysis` - Sales analysis by customer city

**Required Columns:**
- customer_metrics: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
- product_metrics: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
- supplier_metrics: Supplier_ID, Supplier_Name, Product_Count
- regional_analysis: City, Total_Sales, Customer_Count, Avg_Customer_Value

ERROR: Error in parse(text = input): <text>:1:1: unexpected '<'
1: <
    ^


In [ ]:
# REQUIRED variable name: customer_metrics
# Your code here: Calculate customer lifetime value metrics
# Required columns: CustomerID, Name, Total_Spent, Order_Count, Avg_Order_Value
customer_metrics <- complete_data %>%
  group_by(CustomerID, Name) %>%
  summarise(
    Total_Spent = sum(Unit_Price * Quantity, na.rm = TRUE),
    Order_Count = n_distinct(OrderID),
    Avg_Order_Value = ifelse(Order_Count > 0, Total_Spent / Order_Count, 0)
  ) %>%
  ungroup()

# Required output for autograding:
cat("Customer Analysis Summary:
")
cat("Total customers analyzed:", nrow(customer_metrics), "
")
cat("Top customer total spent: $", round(max(customer_metrics$Total_Spent, na.rm = TRUE), 2), "
")
print(head(customer_metrics[order(-customer_metrics$Total_Spent), ], 3))

`summarise()` has grouped output by 'CustomerID'. You can override using the
`.groups` argument.


Customer Analysis Summary:
Total customers analyzed: 94 
Top customer total spent: $ 8471.51 


# A tibble: 3 × 5
  CustomerID Name        Total_Spent Order_Count Avg_Order_Value
       <dbl> <chr>             <dbl>       <int>           <dbl>
1         53 Customer 53       8472.           2           4236.
2          3 Customer 3        6107.           2           3053.
3         61 Customer 61       5768.           2           2884.


In [ ]:
# REQUIRED variable name: product_metrics  
# Your code here: Analyze product performance
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Unit_Price * Quantity, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID)
  ) %>%
  ungroup()

# Required output for autograding:
cat("Product Analysis Summary:
")
cat("Total products analyzed:", nrow(product_metrics), "
")
cat("Top product revenue: $", round(max(product_metrics$Total_Revenue, na.rm = TRUE), 2), "
")
print(head(product_metrics[order(-product_metrics$Total_Revenue), ], 3))

`summarise()` has grouped output by 'ProductID'. You can override using the
`.groups` argument.


Product Analysis Summary:
Total products analyzed: 50 
Top product revenue: $ 11763.16 
# A tibble: 3 × 5
  ProductID Product_Name Total_Revenue Total_Quantity Order_Frequency
      <dbl> <chr>                <dbl>          <dbl>           <int>
1        47 Product 47          11763.             35               8
2        26 Product 26          10854.             42              13
3        43 Product 43          10584.             36              10


In [ ]:
# REQUIRED variable name: supplier_metrics
# Your code here: Evaluate supplier performance  
# Required columns: Supplier_ID, Supplier_Name, Product_Count
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID)
  ) %>%
  ungroup()

# Required output for autograding:
cat("Supplier Analysis Summary:
")
cat("Total suppliers analyzed:", nrow(supplier_metrics), "
")
cat("Max products per supplier:", max(supplier_metrics$Product_Count, na.rm = TRUE), "
")
print(supplier_metrics[order(-supplier_metrics$Product_Count), ])

`summarise()` has grouped output by 'Supplier_ID'. You can override using the
`.groups` argument.


Supplier Analysis Summary:
Total suppliers analyzed: 10 
Max products per supplier: 10 
# A tibble: 10 × 3
   Supplier_ID Supplier_Name Product_Count
         <dbl> <chr>                 <int>
 1           7 Supplier 7               10
 2           4 Supplier 4                6
 3          10 Supplier 10               6
 4           1 Supplier 1                5
 5           3 Supplier 3                5
 6           5 Supplier 5                5
 7           6 Supplier 6                5
 8           2 Supplier 2                3
 9           8 Supplier 8                3
10           9 Supplier 9                2


In [ ]:
# REQUIRED variable name: regional_analysis
# Your code here: Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Unit_Price * Quantity, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)
  ) %>%
  ungroup()

# Required output for autograding:
cat("Regional Analysis Summary:
")
cat("Total cities analyzed:", nrow(regional_analysis), "
")
cat("Highest city sales: $", round(max(regional_analysis$Total_Sales, na.rm = TRUE), 2), "
")
print(regional_analysis[order(-regional_analysis$Total_Sales), ])

Regional Analysis Summary:
Total cities analyzed: 5 
Highest city sales: $ 72277.52 
# A tibble: 5 × 4
  City        Total_Sales Customer_Count Avg_Customer_Value
  <chr>             <dbl>          <int>              <dbl>
1 Chicago          72278.             23              3143.
2 Phoenix          49333.             19              2596.
3 Los Angeles      48506.             18              2695.
4 Houston          38143.             18              2119.
5 New York         36417.             16              2276.


In [ ]:
# Your code here:
# Analyze product performance metrics
# Required columns: ProductID, Product_Name, Total_Revenue, Total_Quantity, Order_Frequency
product_metrics <- complete_data %>%
  group_by(ProductID, Product_Name) %>%
  summarise(
    Total_Revenue = sum(Unit_Price * Quantity, na.rm = TRUE),
    Total_Quantity = sum(Quantity, na.rm = TRUE),
    Order_Frequency = n_distinct(OrderID)
  ) %>%
  ungroup()


`summarise()` has grouped output by 'ProductID'. You can override using the
`.groups` argument.


In [ ]:
# Your code here:
# Evaluate supplier performance
# Required columns: Supplier_ID, Supplier_Name, Product_Count
supplier_metrics <- complete_data %>%
  group_by(Supplier_ID, Supplier_Name) %>%
  summarise(
    Product_Count = n_distinct(ProductID)
  ) %>%
  ungroup()


`summarise()` has grouped output by 'Supplier_ID'. You can override using the
`.groups` argument.


In [ ]:
# Your code here:
# Create regional analysis by customer city
# Required columns: City, Total_Sales, Customer_Count, Avg_Customer_Value
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Unit_Price * Quantity, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Avg_Customer_Value = ifelse(Customer_Count > 0, Total_Sales / Customer_Count, 0)
  ) %>%
  ungroup()


In [ ]:
# Supplier Performance Analysis
supplier_metrics <- complete_data %>%
  filter(!is.na(Supplier_Name)) %>%
  group_by(Supplier_ID, Supplier_Name, Country) %>%
  summarise(
    Total_Revenue = sum(Quantity * Unit_Price, na.rm = TRUE),
    Products_Supplied = n_distinct(ProductID),
    Orders_Involved = n_distinct(OrderID),
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Revenue))

cat("Supplier Performance:\n")
print(supplier_metrics)

Supplier Performance:
# A tibble: 10 × 6
   Supplier_ID Supplier_Name Country     Total_Revenue Products_Supplied
         <dbl> <chr>         <chr>               <dbl>             <int>
 1           5 Supplier 5    Japan              43434.                 5
 2           7 Supplier 7    Germany            35416.                10
 3          10 Supplier 10   Japan              31875.                 6
 4           1 Supplier 1    USA                29706.                 5
 5           3 Supplier 3    Japan              25989.                 5
 6           4 Supplier 4    South Korea        25341.                 6
 7           6 Supplier 6    Japan              21003.                 5
 8           9 Supplier 9    Japan              14845.                 2
 9           2 Supplier 2    USA                10263.                 3
10           8 Supplier 8    Germany             6805.                 3
# ℹ 1 more variable: Orders_Involved <int>


In [ ]:
# Regional Analysis by Customer City
regional_analysis <- complete_data %>%
  group_by(City) %>%
  summarise(
    Total_Sales = sum(Quantity * Unit_Price, na.rm = TRUE),
    Customer_Count = n_distinct(CustomerID),
    Order_Count = n_distinct(OrderID),
    Avg_Customer_Value = Total_Sales / Customer_Count,
    .groups = "drop"
  ) %>%
  arrange(desc(Total_Sales))

cat("Regional Sales Analysis:\n")
print(regional_analysis)

Regional Sales Analysis:
# A tibble: 5 × 5
  City        Total_Sales Customer_Count Order_Count Avg_Customer_Value
  <chr>             <dbl>          <int>       <int>              <dbl>
1 Chicago          72278.             23          36              3143.
2 Phoenix          49333.             19          32              2596.
3 Los Angeles      48506.             18          32              2695.
4 Houston          38143.             18          30              2119.
5 New York         36417.             16          26              2276.


## Part 6: Complex Business Questions

**Tasks:**
Answer the following business questions using your joined datasets:

1. **Customer Segmentation:** Identify your top 10% of customers by total spending
2. **Product Recommendations:** Which products are frequently bought together?
3. **Supplier Dependency:** Which suppliers are most critical to the business?
4. **Market Expansion:** Which cities have high customer counts but low average order values?

Provide data-driven answers with supporting analysis.

## Part 6: Complex Business Questions

**Requirements for Autograding:**
You must create these EXACT variable names:

1. `top_customers` - Top 10% of customers by spending (use quantile function)
2. `product_combinations` - Products frequently bought together  
3. `critical_suppliers` - Suppliers ranked by importance
4. `market_expansion` - Cities with expansion opportunities

**Expected Outputs:**
Each analysis should include specific metrics and rankings as shown in the required outputs below.

In [ ]:
# REQUIRED variable name: top_customers
# Your code here: Identify top 10% of customers by total spending
# Use quantile() function to find the 90th percentile threshold
top_customers <- customer_metrics %>%
  filter(Total_Spent >= quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)) %>%
  arrange(desc(Total_Spent))


# Required output for autograding (uncomment after completing your analysis):
top_10_percent_threshold <- quantile(customer_metrics$Total_Spent, 0.9, na.rm = TRUE)
cat("Customer Segmentation Results:\n")
cat("Top 10% spending threshold: $", round(top_10_percent_threshold, 2), "\n")
cat("Number of top customers:", nrow(top_customers), "\n")

Customer Segmentation Results:
Top 10% spending threshold: $ 4931.51 
Number of top customers: 10 


In [ ]:
# REQUIRED variable name: product_combinations
# Your code here: Find products frequently bought together
# Hint: Group by OrderID, filter orders with multiple items
product_combinations <- order_items %>%
  group_by(OrderID) %>%
  filter(n() > 1) %>%
  summarise(Products = paste(sort(ProductID), collapse = ", ")) %>%
  ungroup() %>%
  group_by(Products) %>%
  summarise(n = n()) %>%
  arrange(desc(n))

# Required output for autograding (uncomment after completing your analysis):
cat("Product Combination Analysis:\n")
cat("Total product combinations found:", nrow(product_combinations), "\n")
if(nrow(product_combinations) > 0) {
cat("Most frequent combination appears", max(product_combinations$n, na.rm = TRUE), "times\n")
}

Product Combination Analysis:
Total product combinations found: 119 
Most frequent combination appears 2 times


In [ ]:
# REQUIRED variable name: critical_suppliers
# Your code here: Analyze which suppliers are most critical
# Consider both revenue contribution and product diversity
critical_suppliers <- supplier_metrics %>%
  filter(Total_Revenue >= quantile(supplier_metrics$Total_Revenue, 0.75, na.rm = TRUE) &
         Products_Supplied >= quantile(supplier_metrics$Products_Supplied, 0.75, na.rm = TRUE)) %>%
  arrange(desc(Total_Revenue), desc(Products_Supplied))

# Required output for autograding (uncomment after completing your analysis):
cat("Critical Suppliers Analysis:\n")
cat("Total suppliers analyzed:", nrow(critical_suppliers), "\n")
print("Top 3 most critical suppliers:")
print(head(critical_suppliers, 3))

Critical Suppliers Analysis:
Total suppliers analyzed: 2 
[1] "Top 3 most critical suppliers:"
# A tibble: 2 × 6
  Supplier_ID Supplier_Name Country Total_Revenue Products_Supplied
        <dbl> <chr>         <chr>           <dbl>             <int>
1           7 Supplier 7    Germany        35416.                10
2          10 Supplier 10   Japan          31875.                 6
# ℹ 1 more variable: Orders_Involved <int>


In [ ]:
# REQUIRED variable name: market_expansion  
# Your code here: Identify market expansion opportunities
# Focus on cities with multiple customers but varying order values
market_expansion <- regional_analysis %>%
  filter(Customer_Count >= quantile(regional_analysis$Customer_Count, 0.75, na.rm = TRUE) &
         Avg_Customer_Value <= quantile(regional_analysis$Avg_Customer_Value, 0.25, na.rm = TRUE)) %>%
  arrange(desc(Customer_Count), Avg_Customer_Value)

# Required output for autograding (uncomment after completing your analysis):
cat("Market Expansion Analysis:\n")
cat("Cities evaluated for expansion:", nrow(market_expansion), "\n")
print("Top expansion opportunities:")
print(head(market_expansion, 5))

Market Expansion Analysis:
Cities evaluated for expansion: 0 
[1] "Top expansion opportunities:"
# A tibble: 0 × 5
# ℹ 5 variables: City <chr>, Total_Sales <dbl>, Customer_Count <int>,
#   Order_Count <int>, Avg_Customer_Value <dbl>


## Part 7: Summary and Insights

**Requirements for Autograding:**
Complete the analysis summary below with specific findings from your work. Your responses will be evaluated for completeness and accuracy.

**Grading Criteria:**
- Specific metrics and numbers from your analysis
- Clear identification of data quality issues  
- Actionable business recommendations
- Technical insights about join performance and appropriateness

## Analysis Summary

### Key Findings from Join Operations:
- **Join Efficiency:** [200 percent of customers matched with orders for inner joins. 77.5 percent of order items retained through all joins.]
- **Data Relationships:** [ orders table had the most orders while customer table had 100 unique customers.]
- **Multi-table Joins:** [When joining multiple tables it started with 400 order items. Then adding products, customers and suppliers it maintained 310 rows.]

### Data Quality Issues Discovered:
- **Orphaned Records:** [50 out of 250 orders without a valid customer. 12.5 percent of all records.]
- **Missing Data:** [Order refrence has non existing customers]
- **Referential Integrity:** [22.5 percent of data was lost due to integrity issues.]

### Business Insights:
- **Top Customer:** [Customer 53 and total amount spent is 8471.51. with 2 orders.]
- **Best Product:** [Product 47 total revenue is 11763.16]
- **Regional Performance:** [citys with higher customer counts usually have higher total sales.]
- **Supplier Analysis:** [10 total suppliers top 3 suppliers control 45.3 percent of revenue.]

### Strategic Recommendations:
- [Maybe expand supplier portfolio consodering the top 3 control so much revenue. I would reccomend expanding operations in the key cities with most potential. and lastly i would reccomend addressing the issue with the missing customners on the orders, this could lead to issues later on.]

### Technical Learnings:
- **Inner vs Left Joins:** [you would use a left join when trying to preserve all records and find gaps. and you would use an inner when you only want to see matched records.]
- **Multi-table Strategy:** [Personally I like to take it one step at a time and make sure each join is working as expected before moving on to the next join.]
- **Performance Considerations:** [rows were lost overall when joining multiple tables. And a reasonable percent of data was lost. But overall the joins were efficient.]

In [ ]:
# Optional: Save your analysis results (not required for autograding)
# Your code here: Save key datasets to CSV files if desired
# write_csv(customer_metrics, "customer_metrics.csv")
# write_csv(product_metrics, "product_metrics.csv")
# write_csv(supplier_metrics, "supplier_metrics.csv")
# write_csv(regional_analysis, "regional_analysis.csv")

# Final summary output for autograding verification:
cat("\n=== HOMEWORK 6 COMPLETION SUMMARY ===\n")
cat("✓ Data Import: 5 datasets loaded\n")
cat("✓ Basic Joins: 4 join types completed\n") 
cat("✓ Multi-table Joins: 4-step progression completed\n")
cat("✓ Data Quality: Anti-joins and semi-joins performed\n")
cat("✓ Business Analysis: Customer, product, supplier, and regional analysis completed\n")
cat("✓ Complex Questions: Advanced business scenarios analyzed\n")
cat("✓ Summary: Comprehensive insights documented\n")
cat("\nAll required variables created for autograding verification.\n")